openedx_tagging/core/tagging/models.py

```diff
@@ -456,3 +490,95 @@ def resync(self) -> bool:
             changed = True
 
         return changed
+
+
+class OpenObjectTag(ObjectTag):
+    """
+    Free-text object tag.
+
+    Only needs a free-text taxonomy and a value to be valid.
```

Comment:
OK, sorry for the slow response, but I'm definitely confused here.

I thought the plan was to have different subclasses of `ObjectTag` based on what type of content object was being tagged:

> Maybe it's ObjectTag that needs the subclassing? It could be made abstract, and have subclasses like CourseTag, BlockTag, etc? And instead of having a completely generic object_id field, we make the object_id return a different real field value in the various subclasses (sort of like what pk does for Django models)?

> But to tag content in the LMS/Studio, we need to enforce ``object_id`` as a CourseKey or UsageKey type. So to do this, we subclass ``ObjectTag``, and use the ``openedx.features.tagging.registry`` to register the subclass, so that it will be picked up when this tagging API creates or resyncs object tags.

And I see that _is_ happening here, but we also have "open" vs. "closed" object tags, so we end up with `OpenCourseObjectTag`, `OpenBlockObjectTag`, `ClosedCourseObjectTag`, `ClosedBlockObjectTag`. 

I thought that the "type" of `ObjectTag` being used would be stored in the ObjectTag row itself, but it seems like it's in `Taxonomy.object_tag_class`. In that case, how could we use, say, a Language (System) Taxonomy to tag both courses and blocks (and anything else that might come up in the future)?

The split between `OpenObjectTag` and `ClosedObjectTag` seems strange to me, and I don't see any requirement for this in the current code. Could we not just have one `ObjectTag` base class with

```
    def _check_tag(self):
        """
        Returns True if this ObjectTag has a valid tag value.
        Subclasses should override this method to perform any additional validation for the particular type of object tag.
        """
        # Open taxonomies don't need an associated tag, but we need a value.
        return bool(self._value) if self.taxonomy.allow_free_text else bool(self.tag_id)
```
and then subclassing is only needed for different types of objects, not different types of taxonomy?

Or is there a more significant implementation difference that I'm not seeing?


I also wonder if there are three types of Taxonomy (at the moment), not two:
* "Free-form Taxonomy" - no `Tag` values exist, and the taxonomy is just whatever `ObjectTag._value`s are in use.
* "Closed Taxonomy" - There is a list of allowed `Tag` values stored in the database, though some ObjectTags may have newer or older tag values, hence they keep a text copy of the tag in `_value`
* "Virtual Taxonomy" which is generated dynamically by python code, e.g. get a list of enabled languages (for "Language" taxonomy), get list of authors, get list of organizations, etc. The `Tag` values don't necessarily need to exist in the database, but they could if that makes the implementation easier rather than more complex. Personally I wonder if storing them in the database requires too much synchronization effort.

Reply:
@bradenmacdonald I'm so confused now. I need more help than just these reviews.

bq. The split between OpenObjectTag and ClosedObjectTag seems strange to me, and I don't see any requirement for this in the current code. Could we not just have one ObjectTag base class with

I follow you here -- it's an artificial distinction that doesn't warrant this complexity.

bq. I thought that the "type" of ObjectTag being used would be stored in the ObjectTag row itself, but it seems like it's in Taxonomy.object_tag_class.

This was my attempt at linking the ObjectTag custom classes to their taxonomies, so that we validate a given ObjectTag for use with a taxonomy. B how is the base `oel_tagging` library supposed to support ObjectTag subclass casting when it doesn't even know about different object types?

bq. I also wonder if there are three types of Taxonomy (at the moment), not two:

Yep -- the "system-defined" taxonomies in the MVP are "virtual taxonomies", and we're not creating Tags for all of them: [ModelObjectTags](https://github.com/open-craft/openedx-learning/pull/3/files#diff-2ecc1274a15be58e16fa617e6e5860e1b2815b6cbeffa200daea9676c07c4bfcR91) keep an association to another Model (like Organization or User). 

But since we're trying to subclass ObjectTag instead of Taxonomy here, the logic for this is all down in the ObjectTag subclasses, and this doesn't seem clean either.

I've scheduled a meeting to discuss, because however we're doing these designs and discussions isn't working :(